src/rgw/rgw_rados.cc

```diff
@@ -4495,12 +4496,17 @@ int RGWRados::copy_obj(RGWObjectCtx& obj_ctx,
       auto& ioctx = ref.pool.ioctx();
       ioctx.locator_set_key(loc.loc);
 
-      ret = rgw_rados_operate(dpp, ioctx, loc.oid, &op, null_yield);
+      ret = ref_io_manager.schedule_io(&ioctx, loc.oid, &op, loc);
       if (ret < 0) {
+        ldpp_dout(dpp, 0) << "ERROR: failed to schedule_io on obj=" << loc.oid << ", the error code = " << ret << dendl;
         goto done_ret;
       }
+    }
 
-      ref_objs.push_back(loc);
+    ret = ref_io_manager.drain_ios();
```

Comment:
Thank you really for such detailed illustrating, that's very helpful for me. At this please, I think it may does not have to add completed to all_results, because all_results is used for cleanup. What's your opinion?

Reply:
this drains the rest of the requests sent by the refcount loop above. if `drain()` returns any errors, we'll need to run that rollback logic below - and that rollback logic should undo any of the successful writes that we drained here too

so yes, i do think we need to add these completions to `all_results` here

Reply:
Yeal, my bad. I revise and commit again, please review when you have time.